### PR TITLE
add enclave loading testing in simulation mode

### PR DESCRIPTION
**Pull request by jinghe-INTC**

### CODE REVIEW

**Comment by jinghe-INTC** [comment]
bors try

**Comment by mingweishih** [comment]
> This test looks good.
> 
> Note: We already run all existing tests (except tests that are not applicable for simulation mode like attestation) in simulation mode.

Based on this, do we still need this test (i.e., explicitly testing the simulation mode on non-simulation build)?

**Comment by anakrish** [comment]
> > This test looks good.
> > Note: We already run all existing tests (except tests that are not applicable for simulation mode like attestation) in simulation mode.
> 
> Based on this, do we still need this test (i.e., explicitly testing the simulation mode on non-simulation build)?

While we run most existing tests with simulation mode, we don't yet have a test that specifically tests simulation mode. It is good to have at least one such test. It does not add too much to the testing burden; while still allowing us to extend the test to include simulation mode specific behavior (lack of exception handling; attestation etc).

**Comment by jinghe-INTC** [comment]
bors try

**Comment by radhikaj** [comment]
@jinghe-INTC please see build error below.
 
Member
Author
jinghe-INTCC:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo -DOE_API_VERSION=2 -DOE_BUILD_UNTRUSTED -DOE_REPO_BRANCH_NAME=\"\" -DOE_REPO_LAST_COMMIT=\"d75db53c883076915bc8babe2adcd7efb0675967\" -I..\3rdparty\mbedtls\mbedtls\include -Ihost -I..\include -Ioutput\include /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 -DOEHOSTMR /showIncludes /Fohost\CMakeFiles\oehostmr.dir\dupenv.c.obj /Fdhost\CMakeFiles\oehostmr.dir\oehostmr.pdb /FS -c ..\host\dupenv.c

[93/541] C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo  -Itests\sim-mode\enc -I..\tests\sim-mode\enc /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 /showIncludes /Fotests\sim-mode\enc\CMakeFiles\enc.dir\enc.c.obj /Fdtests\sim-mode\enc\CMakeFiles\enc.dir\ /FS -c ..\tests\sim-mode\enc\enc.c

FAILED: tests/sim-mode/enc/CMakeFiles/enc.dir/enc.c.obj 

C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo  -Itests\sim-mode\enc -I..\tests\sim-mode\enc /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 /showIncludes /Fotests\sim-mode\enc\CMakeFiles\enc.dir\enc.c.obj /Fdtests\sim-mode\enc\CMakeFiles\enc.dir\ /FS -c ..\tests\sim-mode\enc\enc.c

..\tests\sim-mode\enc\enc.c(4): fatal error C1083: Cannot open include file: 'openenclave/edger8r/enclave.h': No such file or directory

[94/541] C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo -DOE_API_VERSION=2 -DOE_BUILD_UNTRUSTED -DOE_REPO_BRANCH_NAME=\"\" -DOE_REPO_LAST_COMMIT=\"d75db53c883076915bc8babe2adcd7efb0675967\" -I..\3rdparty\mbedtls\mbedtls\include -Ihost -I..\include -Ioutput\include /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 -DOEHOSTMR /showIncludes /Fohost\CMakeFiles\oehostmr.dir\sgx\create.c.obj /Fdhost\CMakeFiles\oehostmr.dir\oehostmr.pdb /FS -c ..\host\sgx\create.c

[95/541] C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo -DOE_API_VERSION=2 -DOE_BUILD_UNTRUSTED -DOE_REPO_BRANCH_NAME=\"\" -DOE_REPO_LAST_COMMIT=\"d75db53c883076915bc8babe2adcd7efb0675967\" -I..\3rdparty\mbedtls\mbedtls\include -Ihost -I..\include -Ioutput\include /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 -DOEHOSTMR /showIncludes /Fohost\CMakeFiles\oehostmr.dir\fopen.c.obj /Fdhost\CMakeFiles\oehostmr.dir\oehostmr.pdb /FS -c ..\host\fopen.c

[96/541] C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo  -Itests\sim-mode\enc -I..\tests\sim-mode\enc /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 /showIncludes /Fotests\sim-mode\enc\CMakeFiles\enc.dir\props.c.obj /Fdtests\sim-mode\enc\CMakeFiles\enc.dir\ /FS -c ..\tests\sim-mode\enc\props.c

FAILED: tests/sim-mode/enc/CMakeFiles/enc.dir/props.c.obj 

C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo  -Itests\sim-mode\enc -I..\tests\sim-mode\enc /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 /showIncludes /Fotests\sim-mode\enc\CMakeFiles\enc.dir\props.c.obj /Fdtests\sim-mode\enc\CMakeFiles\enc.dir\ /FS -c ..\tests\sim-mode\enc\props.c

..\tests\sim-mode\enc\props.c(4): fatal error C1083: Cannot open include file: 'openenclave/enclave.h': No such file or directory

[97/541] C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo -DOE_API_VERSION=2 -DOE_BUILD_UNTRUSTED -DOE_REPO_BRANCH_NAME=\"\" -DOE_REPO_LAST_COMMIT=\"d75db53c883076915bc8babe2adcd7efb0675967\" -I..\3rdparty\mbedtls\mbedtls\include -Ihost -I..\include -Ioutput\include /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 -DOEHOSTMR /showIncludes /Fohost\CMakeFiles\oehostmr.dir\sgx\sgxsign.c.obj /Fdhost\CMakeFiles\oehostmr.dir\oehostmr.pdb /FS -c ..\host\sgx\sgxsign.c

[98/541] cmd.exe /C "cd /D C:\Users\oeadmin\workspace\pipelines\Azure-Windows\build\host && clang -c -O2 -fomit-frame-pointer -m64 -IC:/Users/oeadmin/workspace/pipelines/Azure-Windows/include -IC:/Users/oeadmin/workspace/pipelines/Azure-Windows/build/output/include C:/Users/oeadmin/workspace/pipelines/Azure-Windows/host/sgx/enter.c -o enter.obj"

ninja: build stopped: subcommand failed.

**Comment by anakrish** [comment]
Link to failures: https://oe-jenkins-dev.westeurope.cloudapp.azure.com/blue/organizations/jenkins/pipelines%2FAzure-Windows/detail/Azure-Windows/226/pipeline

File: tests/sim-mode/enc/enc.c [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/edger8r/enclave.h>
```
Comment (anakrish) [inline tests/sim-mode/enc/enc.c:4]: This include line seems to be the one causing Windows compilation errors.

**Comment by anakrish** [comment]
@mingweishih The CI fails due to the following error:

```
 C:\PROGRA~2\MICROS~1\2017\BUILDT~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo  -Itests\sim-mode\enc -I..\tests\sim-mode\enc /DWIN32 /D_WINDOWS /wd4566 /wd4200 /MD /O2 /Ob2 /DNDEBUG   /WX /W3 /showIncludes /Fotests\sim-mode\enc\CMakeFiles\enc.dir\props.c.obj /Fdtests\sim-mode\enc\CMakeFiles\enc.dir\ /FS -c ..\tests\sim-mode\enc\props.c

..\tests\sim-mode\enc\props.c(4): fatal error C1083: Cannot open include file: 'openenclave/enclave.h': No such file or directory
```

Is there something subtle going on with `enclave_includ_directories` on Windows that causes it to omit the  oe include folders?

File: tests/sim-mode/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) Open Enclave SDK contributors.
+# Licensed under the MIT License.
+
+add_subdirectory(host)
+add_subdirectory(enc)
```
Comment (mingweishih) [inline tests/sim-mode/CMakeLists.txt:5]:
```suggestion
if (BUILD_ENCLAVES)
    add_subdirectory(enc)
endif()
```

**Comment by jinghe-INTC** [comment]
bors try

**Comment by mingweishih** [comment]
The issue seems resolved now. Could you squash the comments before we merge? Thanks!

**Comment by radhikaj** [comment]
> The issue seems resolved now. Could you squash the comments before we merge? Thanks!

Thanks for your help @mingweishih

**Comment by radhikaj** [comment]
@mingweishih Can you drop your "Changes Requested"

**Comment by jinghe-INTC** [comment]
bors try

**Comment by mingweishih** [comment]
I'd suggest to resolve the conflicts as follows:
-  Generate a patch based on your first commit
   For example, `git show dd51b74318b8647b732eb60d086735ddc471845a > test.patch`
- Create a clean branch with the same name as the branch of this PR (`add_test_sim-mode`) that sync with the latest master. Be sure to either remove or rename the original one before doing this (e.g., `git branch -m add_test_sim-mode add_test_sim-mode_old`)
- Apply the patch (`git apply test.patch`)
- If see conflicts, remove those conflicting changes in the patch (start with `diff --git`) and manually apply that changes.
- Also apply the change in your second commit
- Commit the code (`git commit -a -s`) and do the force push (`git -f push`).

**Comment by jinghe-INTC** [comment]
> I'd suggest to resolve the conflicts as follows:
> 
> * Generate a patch based on your first commit
>   For example, `git show dd51b74318b8647b732eb60d086735ddc471845a > test.patch`
> * Create a clean branch with the same name as the branch of this PR (`add_test_sim-mode`) that sync with the latest master. Be sure to either remove or rename the original one before doing this (e.g., `git branch -m add_test_sim-mode add_test_sim-mode_old`)
> * Apply the patch (`git apply test.patch`)
> * If see conflicts, remove those conflicting changes in the patch (start with `diff --git`) and manually apply that changes.
> * Also apply the change in your second commit
> * Commit the code (`git commit -a -s`) and do the force push (`git -f push`).

It seems resolved by the recommended steps. Thanks.

**Comment by jinghe-INTC** [comment]
bors try

**Pull request merged**